### ISSUE TITLE
FSDP2 FP8 not giving same results as displayed.

### ISSUE THREAD

**Issue by Manan17** [huggingface/accelerate#3710]
### System Info

```Shell
OS           : Linux-5.15.180.1-1.cm2-x86_64-with-glibc2.35
Python       : 3.10.14
Torch        : 2.7.1+cu126
Accelerate   : 1.9.0

Accelerate Config:

compute_environment: LOCAL_MACHINE
debug: false
distributed_type: FSDP
downcast_bf16: 'no'
enable_cpu_affinity: false
fp8_config:
  amax_compute_algorithm: max
  amax_history_length: 1024
  backend: TE
  fp8_format: HYBRID
  interval: 1
  margin: 0
  override_linear_precision:
  - false
  - false
  - false
  use_autocast_during_eval: false
fsdp_config:
  fsdp_activation_checkpointing: false
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_cpu_ram_efficient_loading: true
  fsdp_offload_params: true
  fsdp_reshard_after_forward: true
  fsdp_state_dict_type: SHARDED_STATE_DICT
  fsdp_transformer_layer_cls_to_wrap: ''
  fsdp_version: 2
machine_rank: 0
main_training_function: main
mixed_precision: fp8
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

Tried to run https://github.com/huggingface/accelerate/blob/main/examples/fsdp2/fsdp2_fp8.py
Did not get the expected results.
Got TFLOPS per device: 50
TPS per device: 1200

### Expected behavior

TFLOPS per device: 450
TPS per device: 10000

**Comment by S1ro1** [comment]
Hi, the example was tested only with fp8 from ao, I see you also apply TE fp8 from the config file. Also your config file specifies that `fsdp_offload_params: true` which will afaik override it, meaning the script will use that. 

The script should be runnable standalone: `torchrun --nproc-per-node=8 fsdp2_fp8.py ...` or `accelerate launch ...` (if you don't have default accelerate config file). Also the behavior is expected, but can be wildly different between versions of torch/torchao from my experience.

**Comment by Manan17** [comment]
I tried removing the accelerate config file, and used accelerate launch, got a TFLOPS of 238 per device.

**Comment by Manan17** [comment]
Also, @S1ro1 can you tell me why does the 1st and the last layer are not in FP8?

**Comment by S1ro1** [comment]
It's to save-up on numerics, with fp8 the accuracy drops significantly. As for the results, 250 sounds like something I've managed to get on one of the setups I tried. There is some change in versions making the training significantly worse (450 -> 250) but I haven't been able to figure out what library and version change causes that. 

In general though, the fp8 should be ~15-20% faster than bf16, which I hope you're seeing.

**Comment by Manan17** [comment]
Can you send me the versions of all the packages you used for this?
I used 0.12.0v for torchao.

**Comment by S1ro1** [comment]
Yeah that's the thing, it's long gone and forgotten. It would probably be something along the newest packages that were public at the point of the PR that landed this, but I constantly upgrade my env so I don't have it anymore. At some point I'll probably just update the docs with new correct numbers...
